src/lib/parser.h

```diff
@@ -73,6 +73,16 @@ static inline void unused(void *dummy, ...) { (void)(dummy);}
     finalize_cmd; \
   } while (0)
 
+/* To save cycles, some places in code makes direct state-transition without
+ * passing through the parser main loop. When switch compilation LIBRDB_DEBUG is
+ * set, DBG_RETURN() enforces return to main loop, and in turn, to print the
+ * state transition */
+#if (LIBRDB_DEBUG)
+    #define DBG_RETURN(cmd) return (cmd)
+#else
+    #define DBG_RETURN(cmd) cmd
+#endif
```

Comment:
I think this is going to hurt readability for the new folks (and for me :smile: ) . It's really difficult to consider both cases: if compiled with a flag it will return, otherwise it will continue. 

Maybe this is not the bottleneck at all, did you benchmark it? Maybe we can go with `return` until we prove it is the bottleneck.

Reply:
i don't like that at all.
i think flow control (return, break, if) are usually very dangerous to hide in macros.
making them optional seems much much worse.

Reply:
The name of the macro actually place upfront exactly what it means which  is RETURN on DBG.

This is very helpfull to investigate corrupted RDB. Maybe the following detailed output that reflects each step transition  might convience you otherwise:

```
➜  push-to-librdb git:(add-stream-support) make noopt > /dev/null                        
➜  push-to-librdb git:(add-stream-support) ./bin/rdb-cli ./test/dumps/stream_v11.rdb resp > /dev/null
➜  push-to-librdb git:(add-stream-support) cat ./rdb-cli.log 
INFO  : RDBX_createReaderFile: Initialized with file ./test/dumps/stream_v11.rdb
INFO  : Finalizing parser configuration
INFO  : Start processing RDB source
INFO  : The parsed RDB file version is: 11
WARN  : Target Redis version is not configured! Set it to Redis version: 7.2
INFO  : rdbversion=11
DEBUG : [0x00000000] elementRdbHeader(State=0)
DEBUG : [0x00000009] elementNextRdbType(State=0) [Opcode=250]
DEBUG : [0x0000000a] elementAuxField(State=0)
DEBUG : [0x0000001c] elementNextRdbType(State=0) [Opcode=250]
DEBUG : [0x0000001d] elementAuxField(State=0)
DEBUG : [0x0000002a] elementNextRdbType(State=0) [Opcode=250]
DEBUG : [0x0000002b] elementAuxField(State=0)
DEBUG : [0x00000036] elementNextRdbType(State=0) [Opcode=250]
DEBUG : [0x00000037] elementAuxField(State=0)
DEBUG : [0x00000045] elementNextRdbType(State=0) [Opcode=250]
DEBUG : [0x00000046] elementAuxField(State=0)
DEBUG : [0x00000051] elementNextRdbType(State=0) [Opcode=254]
DEBUG : [0x00000052] elementSelectDb(State=0) [SelectDB=0]
DEBUG : [0x00000053] elementNextRdbType(State=0) [Opcode=251]
DEBUG : [0x00000054] elementResizeDb(State=0)
DEBUG : [0x00000056] elementNextRdbType(State=0) [Opcode=021]
DEBUG : [0x00000057] elementNewKey(State=0) [key=mystream]
DEBUG : [0x00000060] elementStreamLP(State=0) [key=mystream]
DEBUG : [0x00000061] elementStreamLP(State=1) [key=mystream]
DEBUG : [0x000000fa] elementStreamLP(State=1) [key=mystream]
DEBUG : [0x000000fa] elementStreamLP(State=2) [key=mystream]
DEBUG : [0x00000113] elementStreamLP(State=3) [key=mystream]
DEBUG : [0x00000126] elementStreamLP(State=4) [key=mystream]
DEBUG : [0x0000013f] elementStreamLP(State=4) [key=mystream]
DEBUG : [0x0000013f] elementStreamLP(State=5) [key=mystream]
DEBUG : [0x00000140] elementStreamLP(State=6) [key=mystream]
DEBUG : [0x0000015c] elementStreamLP(State=7) [key=mystream]
DEBUG : [0x0000015c] elementStreamLP(State=6) [key=mystream]
DEBUG : [0x00000178] elementStreamLP(State=7) [key=mystream]
DEBUG : [0x00000188] elementStreamLP(State=7) [key=mystream]
DEBUG : [0x00000188] elementStreamLP(State=3) [key=mystream]
DEBUG : [0x0000019b] elementStreamLP(State=4) [key=mystream]
DEBUG : [0x000001b4] elementStreamLP(State=4) [key=mystream]
DEBUG : [0x000001b4] elementStreamLP(State=5) [key=mystream]
DEBUG : [0x000001b5] elementStreamLP(State=6) [key=mystream]
DEBUG : [0x000001d1] elementStreamLP(State=7) [key=mystream]
DEBUG : [0x000001e1] elementStreamLP(State=7) [key=mystream]
DEBUG : [0x000001e1] elementStreamLP(State=3) [key=mystream]
DEBUG : [0x000001e1] elementEndKey(State=0)
DEBUG : [0x000001e1] elementNextRdbType(State=0) [Opcode=255]
DEBUG : [0x000001e2] elementEndOfFile(State=0) >>> [Status=(ENDED)]
INFO  : Parser done
DEBUG : RDB Reader: Closing file ./test/dumps/stream_v11.rdb        
```

Reply:
I meant you can use `return` unconditionally. So, you can still get debug logs.
I see from the above comment, you added that macro for better performance. I'm just not sure if it really impacts the performance. 

Anyway, leaving decision to you guys.

Reply:
I know that most probably it won't have any impact in isolated test. The reason I feel ambigious about whether to remove this optimization or not is because every step in any given state of the parser can implicitly return due to partial data (i.e. will return `RDB_STATUS_WAIT_MORE_DATA`) without any real issue. This behavior is an intentional design choice to support async parsing.

Reply:
I prefer to drop this macro rather than complicates the readability of this code even more (We still goanna have printouts excepts of the transitions that are being optimized.